test/expr/primary/literal/collection_upcast_opt.swift

```diff
@@ -0,0 +1,38 @@
+// RUN: %target-typecheck-verify-swift -dump-ast 2> %t.ast
+// RUN: %FileCheck %s < %t.ast
+
+// Verify that upcasts of array literals upcast the individual elements in place
+// rather than
+
+protocol P { }
+struct X : P { }
+
+struct TakesArray<T> {
+  init(_: [(T) -> Void]) { }
+}
+
+// CHECK-LABEL: func_decl "arrayUpcast(_:_:)"
+// CHECK: assign_expr
+// CHECK-NOT: collection_upcast_expr
+// CHECK: array_expr type='[(X) -> Void]'
+// CHECK-NEXT: function_conversion_expr implicit type='(X) -> Void'
+// CHECK-NEXT: {{declref_expr.*x1}}
+// CHECK-NEXT: function_conversion_expr implicit type='(X) -> Void'
+// CHECK-NEXT: {{declref_expr.*x2}}
+func arrayUpcast(_ x1: @escaping (P) -> Void, _ x2: @escaping (P) -> Void) {
+  _ = TakesArray<X>([x1, x2])
```

Comment:
The change in behavior here comes from the fact we now open array/dictionary literals early (for performance) which means that we'd attempt `P` of element type and consider everything done. Before we'd attempt only `X` which didn't require conversions. It's good that I actually made conversions like this supported in the solver, but it looks like not completely...